ext/java/org/msgpack/jruby/Encoder.java

```diff
@@ -68,9 +71,15 @@ private void ensureRemainingCapacity(int c) {
   }
 
   private IRubyObject readRubyString() {
-    IRubyObject str = runtime.newString(new ByteList(buffer.array(), 0, buffer.position(), binaryEncoding, false));
-    buffer.clear();
-    return str;
+    if (recursiveExtension) {
+      // if recursiveExtension is true, it means we re-entered encode,
+      // so we MUST NOT flush the buffer.
+      return runtime.newString();
```

Comment:
> Because this method is the only place the buffer is consumed. All other reads call this one.

Ah, I see. I was not sure about it. Could you update this comment to add "All other reads should call this one" or something else?

Reply:
Done.